### PR TITLE
Implement webdav SEARCH

### LINKED ISSUE #676: Search API
Hey guys, for android I would like to implement a real search. Is there already a API for this? It needs to take a search string and returns a list of files available for the user that match the query.

### PR DESCRIPTION
**Pull request by icewind1991**
Still WIP, but submitting so the clients can start with it, functionality wise it should be ready for clients to start using the API

- [x] requires https://github.com/nextcloud/3rdparty/pull/34
- [x] needs proper test coverage

- Adds a mechanism to describe more complex search queries
- Extend the cache to support doing these searches
- Expose the new search mechanism in `\OCP\Files\Folder`
- Implement [rfc5323 SEARCH](https://tools.ietf.org/html/rfc5323) to expose this to the world
- fixes #676

Clients can do a search by firing a SEARCH request at `nextcloud/remote.php/dav` with a search body as described by the rfc

For an example of a somewhat complex query that is possible with this api see [this example request](https://gist.github.com/icewind1991/335d0559cbefa666e7895287189879e7)

cc @mario

### CODE REVIEW

**Comment by icewind1991** [comment]
Fixes #3253

**Review by MorrisJobke — CHANGES_REQUESTED** [review]
Doesn't work:

```
$ curl -X SEARCH http://192.168.99.100/server/remote.php/dav -u admin:admin    
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\NotImplemented</s:exception>
  <s:message>There was no plugin in the system that was willing to handle this SEARCH method.</s:message>
</d:error>
```

**Comment by icewind1991** [comment]
@MorrisJobke you need to send a request body

    curl -u test:test -X SEARCH  https://localcloud.icewind.me/remote.php/dav -H "content-Type: text/xml" --data '<?xml version="1.0"?>
     <d:searchrequest xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns">
     <d:basicsearch>
         <d:select>
             <d:prop>
                 <oc:fileid/>
                 <d:getcontenttype/>
                 <d:getetag/>
                 <oc:size/>
             </d:prop>
         </d:select>
         <d:from>
             <d:scope>
                 <d:href>/files/test</d:href>
                 <d:depth>infinity</d:depth>
             </d:scope>
         </d:from>
         <d:where>
             <d:like>
                 <d:prop>
                     <d:getcontenttype/>
                 </d:prop>
                 <d:literal>text/%</d:literal>
             </d:like>
         </d:where>
         <d:orderby/>
     </d:basicsearch>
  </d:searchrequest>'

**Comment by MorrisJobke** [comment]
I send it but it seems I forgot the text/xml header - let me retry then

File: apps/dav/lib/Files/FileSearchBackend.php [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+		if (!$node instanceof Directory) {
+			throw new \InvalidArgumentException('Search is only supported on directories');
+		}
+
+		$fileInfo = $node->getFileInfo();
+		$folder = $this->rootFolder->get($fileInfo->getPath());
+		/** @var Folder $folder $results */
+		$results = $folder->search($query);
+
+		var_dump(array_map(function (Node $node) {
```
Comment (LukasReschke) [inline apps/dav/lib/Files/FileSearchBackend.php]: expected? 😉

Comment (icewind1991) [inline apps/dav/lib/Files/FileSearchBackend.php]: Leaving debug code in production saves me from having to create a debug patch in the future 😄

**Comment by MorrisJobke** [comment]
If I create a folder `files/test` I only get this as answer:

```xml
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>InvalidArgumentException</s:exception>
  <s:message>Search is only supported on directories</s:message>
</d:error>
```

**Comment by icewind1991** [comment]
@MorrisJobke can you paste the full request you made?

**Comment by MorrisJobke** [comment]
```
$ curl -X SEARCH https://example.org/remote.php/dav -u admin:admin -i -H "content-Type: text/xml" --data '<?xml version="1.0"?> 
 <d:searchrequest xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns">
 <d:basicsearch>
     <d:select>
         <d:prop>
             <oc:fileid/>
             <d:getcontenttype/>
             <d:getetag/>
             <oc:size/>
         </d:prop>
     </d:select>
     <d:from>
         <d:scope>
             <d:href>/files/test</d:href>
             <d:depth>infinity</d:depth>
         </d:scope>
     </d:from>
     <d:where>
         <d:like>
             <d:prop>
                 <d:getcontenttype/>
             </d:prop>
             <d:literal>text/%</d:literal>
         </d:like>
     </d:where>
     <d:orderby/>
</d:basicsearch>
</d:searchrequest>'

HTTP/1.1 404 Not Found
Date: Wed, 15 Feb 2017 16:36:52 GMT
...
Content-Length: 226
Content-Type: application/xml; charset=utf-8

<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\NotFound</s:exception>
  <s:message>Principal with name test not found</s:message>
</d:error>
```

**Comment by MorrisJobke** [comment]
bildschirmfoto 2017-02-15 um 10 40 43

**Comment by icewind1991** [comment]
@MorrisJobke the `scope` paths are relative to the dav root (`remote.php/dav`), so settings the scope to `files/test` searches in the root folder of the `test` user

**Comment by MorrisJobke** [comment]
> @MorrisJobke the scope paths are relative to the dav root (remote.php/dav), so settings the scope to files/test searches in the root folder of the test user

Tested and works 👍 

cc @marinofaggiana I tested it against your test setup and it works 😃

**Comment by MorrisJobke** [comment]
For this body:

```xml
<?xml version="1.0"?><d:searchrequest xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns"><d:basicsearch><d:select><d:prop><d:displayname/></d:prop></d:select><d:from><d:scope><d:href>/files/admin</d:href><d:depth>infinity</d:depth></d:scope></d:from><d:where><d:like><d:prop><d:displayname/></d:prop><d:literal>Documents</d:literal></d:like></d:where></d:basicsearch></d:searchrequest>
```


I get following response:

```xml
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\ServiceUnavailable</s:exception>
  <s:message>TypeError: Argument 4 passed to OC\Files\Search\SearchQuery::__construct() must be of the type array, null given, called in /var/www/nextcloud/apps/dav/lib/Files/FileSearchBackend.php on line 177</s:message>
</d:error>
```

**Comment by LukasReschke** [comment]
>   <s:message>TypeError: Argument 4 passed to OC\Files\Search\SearchQuery::__construct() must be of the type array, null given, called in /var/www/nextcloud/apps/dav/lib/Files/FileSearchBackend.php on line 177</s:message>

General note: Can we avoid leaking error messages? They can sometimes contain sensitive content :)

**Comment by icewind1991** [comment]
@MorrisJobke can you retry with the updated 3rdparty

**Comment by MorrisJobke** [comment]
> @MorrisJobke can you retry with the updated 3rdparty

I now get:

```xml
HTTP/1.1 207 Multi-Status

<?xml version="1.0"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:cal="urn:ietf:params:xml:ns:caldav" xmlns:cs="http://calendarserver.org/ns/" xmlns:card="urn:ietf:params:xml:ns:carddav" xmlns:oc="http://owncloud.org/ns" xmlns:nc="http://nextcloud.org/ns"/>
```

😕

**Comment by icewind1991** [comment]
@MorrisJobke what is the request you're trying?

**Comment by MorrisJobke** [comment]
> @MorrisJobke what is the request you're trying?

https://github.com/nextcloud/server/pull/3360#issuecomment-281534251

**Comment by mario** [comment]
@icewind1991 when doing OPTIONS, search doesn't seem to be among supported methods?

**Comment by icewind1991** [comment]
Works for me:


    curl -I -X OPTIONS  -u test:test https://localcloud.icewind.me/remote.php/dav
    ...
    Allow: OPTIONS, GET, HEAD, DELETE, PROPFIND, PUT, PROPPATCH, COPY, MOVE, REPORT, SEARCH

**Comment by mario** [comment]
Ok will check once again.

**Comment by mario** [comment]
@icewind1991 would something like this work with your implementation, since it doesn't look like the current libs supports that draft entirely:

 <d:searchrequest xmlns:d="DAV:" dcr:="http://www.day.com/jcr/webdav/1.0" >
    <dcr:xpath>//sv:node[@sv:name='myapp:paragraph'][1]</dcr:xpath>
 </d:searchrequest>

cc @AndyScherzinger @tobiasKaminsky 

Otherwise we might look into how we can either extend it use or use (possibly?) another lib just for this particular use case if such a thing exists.

**Comment by MorrisJobke** [comment]
```
There were 4 errors:
1) Test\Files\Cache\CacheTest::testSearchByQuery
Undefined index: storage_mtime
/drone/src/github.com/nextcloud/server/lib/private/Files/Cache/Cache.php:168
...
```

**Comment by icewind1991** [comment]
all rebased, tests green and ready to merge from my POV

Any further polish/fixes can be added later, since this only touches existing stuff minimally it shouldn't break existing stuff.

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
I merged 3rdparty and updated the submodule ;)

**Comment by mario** [comment]
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\NotImplemented</s:exception>
  <s:message>There was no plugin in the system that was willing to handle this SEARCH method.</s:message>
</d:error>

Ideas? :)

**Comment by MorrisJobke** [comment]
> Ideas? :)

How did you send it? Is the content-type header set properly? Is the body correctly formed?

**Comment by mario** [comment]
<?xml version="1.0" encoding="UTF-8"?><searchrequest xmlns="DAV:"><basicsearch><select><allprop/></select><from><scope><href>/files/mario</href><depth>infinity</depth></scope></from><where><like><prop><getcontenttype/></prop><literal>text/%</literal></like></where></basicsearch></searchrequest>

**Comment by mario** [comment]
Ah yes, I remember I do not see SEARCH in allowed methods. Investigating.

**Comment by mario** [comment]
Ok, so two things:

- OPTIONS doesn't respond with SEARCH at all
- difference between /dav vs /webdav? Android seems to be using /webdav and that doesn't work, while /dav works.

**Comment by mario** [comment]
OK so I managed to quickly hack a working basic search grammar so I think I can do this properly as well.

The two questions above still stand though, and also why isn't allprop supported?

cc @icewind1991

**Comment by mario** [comment]
Options started showing SEARCH.

**Comment by MorrisJobke** [comment]
> OPTIONS doesn't respond with SEARCH at all

Needs to be added by @icewind1991 




> difference between /dav vs /webdav? Android seems to be using /webdav and that doesn't work, while /dav works.

remote.php/webdav only contains the webdav of the current user. remote.php/dav contains everything and is also properly namespaced to the user. So `remote.php/webdav/abc` is for the user `admin` the same as `remote.php/dav/files/admin/abc` and for user `test` the same as `remote.php/dav/files/test/abc`. The webdav is the old way and the dav the new way also containing browsable calendar, adressbook, ... ;)

```
dav:/remote.php/dav/> ls
Listing collection `/remote.php/dav/': succeeded.
Coll:   addressbooks                           0  Dez 31  1969
Coll:   calendars                              0  Dez 31  1969
Coll:   comments                               0  Dez 31  1969
Coll:   files                                  0  Dez 31  1969
Coll:   principals                             0  Dez 31  1969
Coll:   public-calendars                       0  Dez 31  1969
Coll:   systemtags-relations                   0  Dez 31  1969
Coll:   systemtags                             0  Dez 31  1969
Coll:   uploads                                0  Dez 31  1969
```

**Comment by mario** [comment]
> OPTIONS doesn't respond with SEARCH at all
> Needs to be added by @icewind1991

When using /dav it responds properly, so no worries.

**Comment by mario** [comment]
Got the DAV vs WEBDAV part explained by @nickvergessen :)

**Comment by marinofaggiana** [comment]
@mario i have lost several days for webdav -> dav 😱

**Comment by mario** [comment]
Hehe. Took me a bit of looking at the differences between mine and Robin
query, but luckily not days hehe :)

**Comment by icewind1991** [comment]
> and also why isn't allprop supported?

Can add `allprop` in a follow up PR, but it shouldn't be used by the clients for search since each requested prop will add a (sometimes significant) cost to the query so only the properties that are actually needed should be requested

**Comment by mario** [comment]
Right, cool @icewind1991 :)

**Comment by icewind1991** [comment]
@MorrisJobke @rullzer @LukasReschke can this be merged?

**Review by MorrisJobke — APPROVED** [review]

**Comment by marinofaggiana** [comment]
Is changed API ??? i have a response error with my driver ...

**Comment by marinofaggiana** [comment]
ping @icewind1991

**Comment by icewind1991** [comment]
@marinofaggiana please describe the error you're getting instead of having us guess

**Comment by marinofaggiana** [comment]
curl -u admin:nextcloudiosapp -X SEARCH https://ios-demo.weasel.rocks/remote.php/dav -H "content-Type: text/xml" --data '<?xml version="1.0"?><d:searchrequest xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns"><d:basicsearch><d:select><d:prop><oc:fileid/><d:getcontenttype/><d:getetag/><oc:size/><oc:favorite/></d:prop></d:select><d:from><d:scope><d:href>/files/admin</d:href><d:depth>infinity</d:depth></d:scope></d:from><d:where><d:like><d:prop><d:displayname/></d:prop><d:literal>Nextcloud%</d:literal></d:like></d:where><d:orderby/></d:basicsearch></d:searchrequest>'

**Comment by marinofaggiana** [comment]
curl -u admin:XXXXXXXX -X SEARCH https://ios-demo.weasel.rocks/remote.php/dav -H "content-Type: text/xml" --data '<?xml version="1.0"?><d:searchrequest xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns"><d:basicsearch><d:select><d:prop><oc:fileid/><d:getcontenttype/><d:getetag/><oc:size/><oc:favorite/></d:prop></d:select><d:from><d:scope><d:href>/files/admin</d:href><d:depth>infinity</d:depth></d:scope></d:from><d:where><d:like><d:prop><d:displayname/></d:prop><d:literal>Nextcloud%</d:literal></d:like></d:where><d:orderby/></d:basicsearch></d:searchrequest>'

**Comment by marinofaggiana** [comment]
@icewind1991 this is the response :  

<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\NotImplemented</s:exception>
  <s:message>There was no plugin in the system that was willing to handle this SEARCH method.</s:message>
</d:error>

**Comment by marinofaggiana** [comment]
@icewind1991 @MorrisJobke please info for this issue.

**Comment by mario** [comment]
So search basically works, so please +1 all of you.

After that, I just need it extended to support oc:favorite in a query because filter-files won't work for Android atm.

**Comment by marinofaggiana** [comment]
@mario 
<?xml version=\"1.0\"?><oc:filter-files xmlns:d=\"DAV:\" xmlns:oc=\"http://owncloud.org/ns\" xmlns:nc=\"http://nextcloud.org/ns\"><oc:filter-rules><oc:favorite>1</oc:favorite></oc:filter-rules><d:prop>

**Comment by mario** [comment]
@marinofaggiana using that our Android lib doesn't want to parse the propstat so I don't get the properties back.

**Comment by marinofaggiana** [comment]
Modify the lib ?

**Comment by mario** [comment]
Not something I feel comfortable doing under these kind of deadlines.

**Comment by MorrisJobke** [comment]
Sorry - I screwed up the instance. It should work now. :( @marinofaggiana @icewind1991

**Review by rullzer — APPROVED** [review]

**Comment by barrydegraaff** [comment]
@mario good work. I have been playing around with WebDAV SEARCH on nextcloud 12.0.4 and integrate it in Zimbra Nextcloud Client (https://github.com/Zimbra-Community/owncloud-zimlet)

Mostly I can make it work, but I have issues searching for files with more than 2 search words.
So I would like to search for all files that match: word1, word2 and word3. So far I can only make Nextcloud look for word1 and word2. The 3rd word is ignored. Where can I find good documentation on webdav search?

       curl -u user:password -X SEARCH https://nextcloud.example.com/nextcloud/remote.php/dav -H "content-Type: text/xml" --data '<d:searchrequest xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns"><d:basicsearch><d:select><d:prop>oc:fileid/<d:getcontentlength/><d:getlastmodified/><d:getcontenttype/><d:getetag/>oc:size/oc:favorite/</d:prop></d:select><d:from><d:scope><d:href>/files/user</d:href><d:depth>infinity</d:depth></d:scope></d:from><d:where><d:and><d:like><d:prop><d:displayname/></d:prop><d:literal>%word1%</d:literal></d:like><d:like><d:prop><d:displayname/></d:prop><d:literal>%word2%</d:literal></d:like><d:like><d:prop><d:displayname/></d:prop><d:literal>%word3%</d:literal></d:like></d:and></d:where><d:orderby><d:order><d:prop><d:displayname/></d:prop><d:ascending/></d:order><d:order><d:prop><d:getlastmodified/></d:prop><d:ascending/></d:order></d:orderby></d:basicsearch></d:searchrequest>'
 
 
        curl -u user:password -X SEARCH https://nextcloud.domain.com/nextcloud/remote.php/dav -H "content-Type: text/xml" --data '<d:searchrequest xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns"><d:basicsearch><d:select><d:prop>oc:fileid/<d:getcontentlength/><d:getlastmodified/><d:getcontenttype/><d:getetag/>oc:size/oc:favorite/</d:prop></d:select><d:from><d:scope><d:href>/files/user</d:href><d:depth>infinity</d:depth></d:scope></d:from><d:where><d:and><d:like><d:prop><d:displayname/></d:prop><d:literal>%word1%</d:literal></d:like><d:like><d:prop><d:displayname/></d:prop><d:literal>%word2%</d:literal></d:like></d:and></d:where><d:orderby><d:order><d:prop><d:displayname/></d:prop><d:ascending/></d:order><d:order><d:prop><d:getlastmodified/></d:prop><d:ascending/></d:order></d:orderby></d:basicsearch></d:searchrequest>'

So here only the first 2 words are matched, the 3rd is ignored... is there a limit of 2 in AND?
      <d:where><d:and><d:like><d:prop><d:displayname/></d:prop><d:literal>%word1%</d:literal></d:like><d:like><d:prop><d:displayname/></d:prop><d:literal>%word2%</d:literal></d:like><d:like><d:prop><d:displayname/></d:prop><d:literal>%word3%</d:literal></d:like></d:and></d:where>

**Comment by barrydegraaff** [comment]
https://www.ietf.org/rfc/rfc5323.txt
5.7.  Boolean Operators: DAV:and, DAV:or, and DAV:not

   The DAV:and operator performs a logical AND operation on the
   expressions it contains.

Can anyone help me to do a WebDAV Search to nextcloud with more than 2 words to look for in displayname, I think the XML example, is according to the spec, but maybe I misread.

      <d:where><d:and><d:like><d:prop><d:displayname/></d:prop><d:literal>%word1%</d:literal></d:like><d:like><d:prop><d:displayname/></d:prop><d:literal>%word2%</d:literal></d:like><d:like><d:prop><d:displayname/></d:prop><d:literal>%word3%</d:literal></d:like></d:and></d:where>

**Comment by icewind1991** [comment]
How I understood the spec (and implemented it) is that DAV:and and DAV:or only take 2 arguments,
you can nest multiple and/or statements to get your desired effect

**Comment by barrydegraaff** [comment]
@icewind1991 thanks, that makes it a bit harder to implement on the client side. It would be nice if the rfc said `The operator takes two arguments.` instead of `The DAV:and operator performs a logical AND operation on the expressions it contains.` I would not read 2 in that.

**Comment by barrydegraaff** [comment]
@icewind1991 I am thinking about asking julian reschke (his name is on the rfc) and see what he says.

But should the server return a 501 Not Implemented error if I do something that is outside the webdav spec? Aka an AND with 3 criteria = 501. Now it parses only the first 2.

**Comment by reschke** [comment]
I agree that RFC 5323 could be clearer, but AFAIR, we would have said "two" if we had intended to restrict it to two operands.

**Comment by icewind1991** [comment]
I will see if I can extend the implementation to allow for arbitrary number of arguments

**Comment by barrydegraaff** [comment]
@icewind1991 thanks!

**Comment by rullzer** [comment]
I just noticed this as well.

Since and is associative you can just do 

```a AND b AND c AND d```

and transform it into 

```((a AND b) AND c) AND d)```

**Comment by icewind1991** [comment]
Fix is [here](https://github.com/nextcloud/server/pull/7882)

**Comment by pinkylabs** [comment]
The [example from @MorrisJobke ](https://github.com/nextcloud/server/pull/3360#issuecomment-280063038) does not work if 'where' element is omitted.

This works OK:
```
<?xml version="1.0" encoding="utf-16"?>
<searchrequest xmlns="DAV:">
  <basicsearch>
    <select>
      <prop>
        <getcontenttype />
        <getetag />
      </prop>
    </select>
    <from>
      <scope>
        <href>/files/pinky/tmp</href>
        <depth>infinity</depth>
      </scope>
    </from>
    <where>
      <like>
        <prop>
          <getcontenttype />
        </prop>
        <literal>text/%</literal>
      </like>
    </where>
  </basicsearch>
</searchrequest>
```
This fails with error 400
```
<?xml version="1.0" encoding="utf-16"?>
<searchrequest xmlns="DAV:">
  <basicsearch>
    <select>
      <prop>
        <getcontenttype />
        <getetag />
      </prop>
    </select>
    <from>
      <scope>
        <href>/files/pinky/tmp</href>
        <depth>infinity</depth>
      </scope>
    </from>
  </basicsearch>
</searchrequest>
```

As far as I understand DTD, the only mandatory elements are select and from: 
```
<!ELEMENT basicsearch   (select, from, where?, orderby?, limit?) >
```
See [section 5.2](https://tools.ietf.org/html/rfc5323#section-5.2)

**Comment by rullzer** [comment]
@pinkylabs please open a new issue.

**Pull request merged**